### PR TITLE
Implement BigInt asIntN and asUintN methods

### PR DESCRIPTION
**Pull request by batizdaniel**
The following methods were implemented:
- BigInt.asIntN
- BigInt.asUintN

Custom dispatcher also added to builtin_bigint.

JerryScript-DCO-1.0-Signed-off-by: Daniel Batiz batizjob@gmail.com

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 1]
```diff
@@ -36,6 +52,358 @@
...
+  if (input_bits == 0 || bigint_value == ECMA_BIGINT_ZERO)
+  {
+    ecma_free_value (bigint_value);
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t four_byte = (sizeof (uint32_t) * sizeof (uint64_t));
+  const uint32_t one_byte = sizeof (uint64_t);
+
+  uint32_t whole_part = (ecma_value_t) input_bits / four_byte;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:96]: input_bits is whole number between [0- 2^53 - 1]. Division by  `four_byte` results a number between [0- 2^48 - 1] which cannot fit into an uint32_t, and also the ecma_value_t cast is invalid.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/* Copyright JS Foundation and other contributors, http://js.foundation
+ /* Copyright JS Foundation and other contributors, http://js.foundation
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:1]: No need this space.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 3]
```diff
@@ -71,6 +439,44 @@ ecma_builtin_bigint_dispatch_construct (const ecma_value_t *arguments_list_p, /*
...
+ecma_builtin_bigint_dispatch_routine (uint8_t builtin_routine_id, /**< built-in wide routine identifier */
+                                      ecma_value_t this_arg, /**< 'this' argument value */
+                                      const ecma_value_t arguments_list_p[], /**< list of arguments
+                                                                              *   passed to routine */
+                                      uint32_t arguments_number) /**< length of arguments' list */
+{
+  JERRY_UNUSED (this_arg);
+
+  ecma_value_t arg1 = (arguments_number > 0) ? arguments_list_p[0] : ECMA_VALUE_UNDEFINED;
+  ecma_value_t arg2 = (arguments_number > 1) ? arguments_list_p[1] : ECMA_VALUE_UNDEFINED;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:458]: according to #4741 these checks are not necessary.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 4]
```diff
@@ -71,6 +439,44 @@ ecma_builtin_bigint_dispatch_construct (const ecma_value_t *arguments_list_p, /*
...
+  }
+
+  switch (builtin_routine_id)
+  {
+    case ECMA_BUILTIN_BIGINT_AS_INT_N:
+    case ECMA_BUILTIN_BIGINT_AS_U_INT_N:
+    {
+      bool is_signed = (builtin_routine_id == ECMA_BUILTIN_BIGINT_AS_INT_N);
+      return ecma_builtin_bigint_object_as_int_n (this_arg, arg1, arg2, is_signed);
+    }
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:472]:
```c
    case ECMA_BUILTIN_BIGINT_AS_INT_N:
    {
      return ecma_builtin_bigint_object_as_int_n (this_arg, arg1, arg2, true);
    }
    case ECMA_BUILTIN_BIGINT_AS_U_INT_N:
    {
      return ecma_builtin_bigint_object_as_int_n (this_arg, arg1, arg2, false);
    }
```

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 5]
```diff
@@ -36,6 +52,358 @@
...
+  }
+
+  if (input_bits == 0 || bigint_value == ECMA_BIGINT_ZERO)
+  {
+    ecma_free_value (bigint_value);
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t four_byte = (sizeof (uint32_t) * sizeof (uint64_t));
+  const uint32_t one_byte = sizeof (uint64_t);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:94]: Is this a bits_in_byte thing? Even though `sizeof (uint64_t)` == 8, this seems an incorrect definition. Same as above, which looks like a bits_in_digit.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 6]
```diff
@@ -36,6 +52,358 @@
...
+  uint32_t input_byte_size = (ecma_value_t) input_bits / one_byte;
+
+  if ((ecma_value_t) input_bits % one_byte != 0)
+  {
+    input_byte_size += 1;
+  }
+
+  ecma_extended_primitive_t * input_bigint_p = ecma_get_extended_primitive_from_value (bigint_value);
+
+  if (JERRY_UNLIKELY (input_bigint_p == NULL))
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:119]: This looks an invalud return. Probably this could be an assert.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 7]
```diff
@@ -36,6 +52,358 @@
...
+  }
+
+  uint32_t input_byte_size = (ecma_value_t) input_bits / one_byte;
+
+  if ((ecma_value_t) input_bits % one_byte != 0)
+  {
+    input_byte_size += 1;
+  }
+
+  ecma_extended_primitive_t * input_bigint_p = ecma_get_extended_primitive_from_value (bigint_value);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:117]: No space after `*`

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 8]
```diff
@@ -36,6 +52,358 @@
...
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_bigint_object_as_int_n (ecma_value_t this_arg, /**< 'this' argument */
+                                     ecma_value_t bits, /**< number of bits */
+                                     ecma_value_t bigint, /**< bigint number */
+                                     bool is_signed) /**< The operation is signed */
+{
+  JERRY_UNUSED (this_arg);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:70]: due to custom dispatcher passing this_args in unnecessary.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 9]
```diff
@@ -36,6 +52,358 @@
...
+  {
+    min_size = sizeof (uint32_t);
+  }
+
+  ecma_extended_primitive_t * result_p = ecma_bigint_create (min_size);
+
+  if (JERRY_UNLIKELY (result_p == NULL))
+  {
+    ecma_deref_bigint (input_bigint_p);
+    return 0;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:144]: This is also suspcious

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 5 above)
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:94]: fix the naming of these variables.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 10]
```diff
@@ -36,6 +52,358 @@
...
+
+  if (JERRY_UNLIKELY (input_bigint_p == NULL))
+  {
+    return 0;
+  }
+
+  uint32_t bigint_size = ECMA_BIGINT_GET_SIZE (input_bigint_p);
+  uint8_t input_bigint_sign = input_bigint_p->u.bigint_sign_and_size & ECMA_BIGINT_SIGN;
+
+  uint32_t min_size = ((input_bit_length / one_byte) < bigint_size) ? (input_bit_length / one_byte) : bigint_size;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:127]: `input_bit_length / one_byte` used many times, maybe we could put it into a local valriable. also this could be a `>>` shift

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 11]
```diff
@@ -36,6 +52,358 @@
...
+  if (JERRY_UNLIKELY (result_p == NULL))
+  {
+    ecma_deref_bigint (input_bigint_p);
+    return 0;
+  }
+
+  ecma_bigint_digit_t *first_digit_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, 0);
+  ecma_bigint_digit_t *last_digit_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, bigint_size);
+
+  /* Calc the leading zeros of the input_bigint */
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:150]: Calc -> calculate. Also add a full stop

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:96]: uint64_t whole_part = floor (input_bits / 32);

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 12]
```diff
@@ -36,6 +52,358 @@
...
+
+  uint32_t bits_of_bigint = (uint32_t) (bigint_size * one_byte) - zeros;
+
+  uint32_t exact_size = 1;
+
+  if ((bits_of_bigint > (four_byte - 1)) || input_bigint_sign)
+  {
+    exact_size = (input_byte_size < bigint_size) ? input_byte_size : bigint_size;
+
+    if ((input_bigint_sign) && (input_byte_size > bigint_size))
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:168]: no need () around input_bigint_sign

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 13]
```diff
@@ -36,6 +52,358 @@
...
+  {
+    ecma_free_value (bigint_value);
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t four_byte = (sizeof (uint32_t) * sizeof (uint64_t));
+  const uint32_t one_byte = sizeof (uint64_t);
+
+  uint32_t whole_part = (ecma_value_t) input_bits / four_byte;
+  uint32_t remainder = (ecma_value_t) input_bits % four_byte;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:97]: uint32_t remainder = (uint32_t) fmod (input_bits, 32);

**Review by zherczeg — CHANGES_REQUESTED** [review]
The code is coming from the standard?

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 14]
```diff
@@ -36,6 +52,358 @@
...
+
+  uint32_t input_bit_length = 1;
+
+  if (whole_part == 0)
+  {
+    input_bit_length = (remainder > 0) ? (ecma_value_t) input_bits : four_byte;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:103]: input_bit_length = (uint64_t) input_bits

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 15]
```diff
@@ -36,6 +52,358 @@
...
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t four_byte = (sizeof (uint32_t) * sizeof (uint64_t));
+  const uint32_t one_byte = sizeof (uint64_t);
+
+  uint32_t whole_part = (ecma_value_t) input_bits / four_byte;
+  uint32_t remainder = (ecma_value_t) input_bits % four_byte;
+
+  uint32_t input_bit_length = 1;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:99]: uint64_t

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 16]
```diff
@@ -36,6 +52,358 @@
...
+
+  if (whole_part == 0)
+  {
+    input_bit_length = (remainder > 0) ? (ecma_value_t) input_bits : four_byte;
+  }
+  else
+  {
+    input_bit_length = (remainder > 0) ? (uint32_t) (whole_part + 1) * four_byte : (uint32_t) (whole_part * four_byte);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:107]: u64 casts

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 17]
```diff
@@ -36,6 +52,358 @@
...
+  }
+
+  uint32_t input_byte_size = (ecma_value_t) input_bits / one_byte;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:110]: u64, floor

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 18]
```diff
@@ -36,6 +52,358 @@
...
+
+  if ((ecma_value_t) input_bits % one_byte != 0)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:112]: fmod

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 6 above)
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:119]: this checks is unnecessary, since it never return non null ptr

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 10 above)
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:127]: u64

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 19]
```diff
@@ -36,6 +52,358 @@
...
+  }
+
+  uint32_t bigint_size = ECMA_BIGINT_GET_SIZE (input_bigint_p);
+  uint8_t input_bigint_sign = input_bigint_p->u.bigint_sign_and_size & ECMA_BIGINT_SIGN;
+
+  uint32_t min_size = ((input_bit_length / one_byte) < bigint_size) ? (input_bit_length / one_byte) : bigint_size;
+
+  if (input_bigint_sign && (input_byte_size > bigint_size))
+  {
+    min_size = ((input_bit_length / one_byte) > bigint_size) ? (input_bit_length / one_byte) : bigint_size;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:131]: no duplicate division operation.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 9 above)
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:144]: return ECMA_VALUE_ERROR

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 20]
```diff
@@ -36,6 +52,358 @@
...
+    uint32_t mask = (uint32_t) 1 << mask_bit;
+
+    if (mask == 0)
+    {
+      mask = ~((ecma_bigint_digit_t) 0);
+    }
+    else if (is_signed && mask > 1)
+    {
+      mask -= 1;
+    }
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:213]: this part in unnecessary.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 21]
```diff
@@ -36,6 +52,358 @@
...
+    if (is_signed)
+    {
+      if ((input_bigint_sign) && ((~digits_p[last_cell] + 1) & check_sign_mask) == 0)
+      {
+        positive_num = 1;
+      }
+
+      uint32_t num_representation = 0;
+
+      if (((digits_p[last_cell]) & check_sign_mask) == 0)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:232]: too many parenthesis.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 22]
```diff
@@ -36,6 +52,358 @@
...
+    int16_t equal_bit_s = 0;
+
+    if (remainder != 0)
+    {
+      equal_bit_s = -1;
+    }
+
+    int32_t last_cell_negative = (last_cell != 0) ? ((int32_t) last_cell * (-1)) : -1;
+
+    uint8_t helper_value = 0;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:263]: find a better name for it.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 23]
```diff
@@ -36,6 +52,358 @@
...
+        }
+        while (first_cell != equal_bit_s);
+
+        first_cell = last_cell_negative;
+      }
+    }
+
+    result_number_p[-1] &= mask;
+
+    uint32_t surplus = (min_size - exact_size) / (sizeof (uint32_t) / sizeof (ecma_char_t));
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:297]: save it into a constans sizeof `sizeof (uint32_t) / sizeof (ecma_char_t))`

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 24]
```diff
@@ -36,6 +52,358 @@
...
+
+    if (min_size / sizeof (uint64_t) < 1)
+    {
+      surplus = 0;
+    }
+
+    if (is_signed)
+    {
+      if ((result_p->u.bigint_sign_and_size) > exact_size && (min_size > sizeof (uint32_t))
+           && result_number_p[last_cell_negative] < 1)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:318]:
```c
if (result_p->u.bigint_sign_and_size > exact_size
    && min_size > sizeof (uint32_t)
    && result_number_p[last_cell_negative] < 1)
```

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 25]
```diff
@@ -36,6 +52,358 @@
...
+    if (new_size < result_p->u.bigint_sign_and_size)
+    {
+      result_p->refs_and_type = ECMA_EXTENDED_PRIMITIVE_REF_ONE | ECMA_TYPE_BIGINT;
+
+      uint32_t new_size_remainder = new_size % sizeof (uint32_t);
+
+      ecma_extended_primitive_t * new_result_p = ecma_bigint_create (new_size - new_size_remainder);
+
+      new_result_p->u.bigint_sign_and_size += new_size_remainder;
+      memcpy (new_result_p +1, result_p +1, new_size - new_size_remainder);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:379]: spaces after `+`

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 26]
```diff
@@ -36,6 +53,353 @@
...
+  if (input_bits == 0 || bigint_value == ECMA_BIGINT_ZERO)
+  {
+    ecma_free_value (bigint_value);
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t size_of_divisor_in_bits = (sizeof (uint32_t) * sizeof (uint64_t));
+
+  uint64_t whole_part = (uint64_t) floor (input_bits / size_of_divisor_in_bits);
+  uint32_t remainder = (uint32_t) fmod (input_bits, size_of_divisor_in_bits);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:97]: The input_bits should be an integer number. Furthermore JerryScript has a maximum length for bigints, any number greater than that does not need any clamping (just increase its refcount and return with it).

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 27]
```diff
@@ -36,6 +53,353 @@
...
+    return bigint_value;
+  }
+
+  if (input_bits == 0 || bigint_value == ECMA_BIGINT_ZERO)
+  {
+    ecma_free_value (bigint_value);
+    return ECMA_BIGINT_ZERO;
+  }
+
+  const uint32_t size_of_divisor_in_bits = (sizeof (uint32_t) * sizeof (uint64_t));
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:94]: This `sizeof (uint64_t)` is still misleading. I suspect you want bits in byte, not the size of an uint64 value.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 28]
```diff
@@ -36,6 +53,353 @@
...
+
+  uint64_t whole_part = (uint64_t) floor (input_bits / size_of_divisor_in_bits);
+  uint32_t remainder = (uint32_t) fmod (input_bits, size_of_divisor_in_bits);
+
+  uint64_t input_bit_length = 1;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:99]: uint32_t should be enough for any valid number in JerryScript.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 29]
```diff
@@ -36,6 +53,353 @@
...
+
+  if (whole_part == 0)
+  {
+    input_bit_length = (uint64_t) input_bits;
+  }
+  else
+  {
+    input_bit_length = (remainder > 0) ? (uint64_t) (whole_part + 1) * size_of_divisor_in_bits
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:107]: ???? This looks like getting the original input size. You have this value, no need for recomputing it.

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 30]
```diff
@@ -36,6 +53,353 @@
...
+    : (uint64_t) (whole_part * size_of_divisor_in_bits);
+  }
+
+  uint64_t input_byte_size = (uint64_t) floor (input_bits / sizeof (uint64_t));
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:111]: The algorith should work with digits, not bytes. This is important for handling big endian systems

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 31]
```diff
@@ -36,6 +53,353 @@
...
+
+  /* Calculate the leading zeros of the input_bigint */
+  ecma_bigint_digit_t zeros;
+
+  do
+  {
+    zeros = ecma_big_uint_count_leading_zero (*first_digit_p);
+    ++first_digit_p;
+  }
+  while (first_digit_p < last_digit_p);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:156]: What is the exact purpose of this while? Isn't it the same as `zeros = ecma_big_uint_count_leading_zero (last_digit_p[-1]);`

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 32]
```diff
@@ -36,6 +53,353 @@
...
+
+    uint32_t mask = (uint32_t) 1 << mask_bit;
+
+    if (mask == 0)
+    {
+      mask = ~((ecma_bigint_digit_t) 0);
+    }
+    else
+    {
+      mask -= 1;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:211]: If mask is 0, `mask - 1` is the same as `~((ecma_bigint_digit_t) 0)`

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 33]
```diff
@@ -36,6 +53,353 @@
...
+    }
+
+    uint32_t last_cell = (exact_size >= sizeof (uint32_t)) ? (uint32_t) (min_size / sizeof (uint32_t)) - 1 : 0;
+
+    uint8_t positive_num = 0;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:216]: Isn't this a bool?

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c [hunk 34]
```diff
@@ -36,6 +53,353 @@
...
+      uint8_t num_representation = 0;
+
+      if ((digits_p[last_cell] & check_sign_mask) == 0)
+      {
+        num_representation = 1;
+      }
+
+      do
+      {
+        *result_number_p++ = num_representation ? *digits_p++ : ~(*digits_p++);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:234]: So the algorithm does a proper conversion, then turns the result back to negative if needed? Can it be done with a single pass?

**Review by zherczeg — CHANGES_REQUESTED** [review]
I would like to get a bit more info about the high level concept of the algorith, so I could suggest optimizations.

**Comment by zherczeg** [comment]
I would also like to have some tests on JerryScript, not just the test262 ones. These tests should check corner cases of the algorithm, not covered by test262

File: jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c (hunk 26 above)
Comment (batizdaniel) [reply jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c:97]: The head of the code is coming from the standard. So the input_bits should be a number according to the ToIndex method.
(https://tc39.es/ecma262/#sec-bigint.asintn)

**Comment by ossy-szeged** [comment]
closing, continue in PR #5165
